src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java

```diff
@@ -0,0 +1,691 @@
+        // third shard which was not loaded before, that third shard might become now eligible.
+        browseForPotentialNewShards();
+        browseForPotentialNewShardsFromAggressiveRelations();
+        // Repeat the same process as long as we find some of those third party shards.
+        while (!this.loadedShards.keySet().equals(currentShards))
+        {
+            if (logger.isInfoEnabled())
+            {
+                final Set<Shard> missingShards = new HashSet<>(this.loadedShards.keySet());
+                missingShards.removeAll(currentShards);
+                logger.info("{}: Preemptive load found new unexpected 2nd degree shard(s): {}",
+                        this.dynamicAtlas.getName(),
+                        missingShards.stream().map(Shard::getName).collect(Collectors.toList()));
+            }
+
+            // Load all the shards into a multiAtlas
+            buildUnderlyingMultiAtlas();
+            // Record the current list of shards
+            currentShards = new HashSet<>(this.loadedShards.keySet());
+            // Loop through the entities again to find potential shards to add.
+            browseForPotentialNewShards();
+            browseForPotentialNewShardsFromAggressiveRelations();
+        }
+        this.preemptiveLoadDone = true;
+    }
+
+    boolean relationCovered(final Relation relation)
+    {
+        final Set<Long> parentRelationIdentifierTree = new HashSet<>();
+        parentRelationIdentifierTree.add(relation.getIdentifier());
+        return relationCoveredInternal(relation, parentRelationIdentifierTree);
+    }
+
+    private void addNewShardLog(final Shard shard)
+    {
+        if (logger.isInfoEnabled())
+        {
+            final Atlas loaded = this.loadedShards.get(shard);
+            if (loaded == null)
+            {
+                logger.info("{}: Loading new shard {} found no new Atlas.",
+                        this.dynamicAtlas.getName(), shard.getName());
+            }
+            else
+            {
+                logger.info("{}: Loading new shard {} found a new Atlas {} of size {}",
+                        this.dynamicAtlas.getName(), shard.getName(), loaded.getName(),
+                        loaded.size());
+            }
+        }
+    }
+
+    private void addNewShards(final Iterable<? extends Shard> shards)
+    {
+        final Set<Shard> initialNonEmptyLoadedShards = nonNullShards();
+        for (final Shard shard : shards)
+        {
+            if (!this.loadedShards.containsKey(shard))
+            {
+                this.loadedShards.put(shard, this.atlasFetcher.apply(shard).orElse(null));
+                addNewShardLog(shard);
+            }
+        }
+        final List<Atlas> nonNullAtlasShards = getNonNullAtlasShards();
+        if (!nonNullAtlasShards.isEmpty())
+        {
+            if (shouldBuildUnderlyingMultiAtlasWhenAddingNewShards(initialNonEmptyLoadedShards))
```

Comment:
Here some complexity has been pushed to this method: `shouldBuildUnderlyingMultiAtlasWhenAddingNewShards `